### PR TITLE
Removed ScheduleManager and ScheduleManagerCore from container runtime

### PR DESCRIPTION
**Pull request by agarwal-navin**
## Description
`ScheduleManager` and `ScheduleManagerCore` are no longer needed. They have two responsibilities as per the comment on the class:
```
/**
 * This class has the following responsibilities:
 *
 * 1. It tracks batches as we process ops and raises "batchBegin" and "batchEnd" events.
 * As part of it, it validates batch correctness (i.e. no system ops in the middle of batch)
 *
 * 2. It creates instance of ScheduleManagerCore that ensures we never start processing ops from batch
 * unless all ops of the batch are in.
 */
```

We don't need `ScheduleManager` to do these things anymore. Batching was complicated before because the Container / Deltamanager were in charge of deciding when a batch would be flushed. However, now the runtime is in control of sending a batch together and then processing it together.

This PR removes both these classes and incorporates the functionality into the processing pipeline directly as follows:

1. Schedule manager tracked batches as we process ops and raises "batchBegin" and "batchEnd" events. As part of it, it validates batch correctness (i.e. no system ops in the middle of batch) - These events are emitted directly in `ContainrRuntime::process` which has knowledge of the entire batch. The batch correctness logic is moved into `RemoteMessageProcessor::process` which already had some of these validations in place.

2. It creates instance of ScheduleManagerCore that ensures we never start processing ops from batch unless all ops of the batch are in - This functionality is already built in into the `RemoteMessageProcessor::process` which ensures that we don't start processing batches until we get the entire batch.

3. `ScheduleManager` called `DeltaScheduler` on batch start and batch end which evaluated if it should yield a JS turn before processing more ops. This functionality is preserved and the container runtime calls `DeltaScheduler` directly now.

## Testing
- The existing tests for batch validation pass. The only change is the error strings are different for some of them.
- The delta scheduler tests were in `packages/test/functional-tests/src/test/containerRuntime.spec.ts`. These tests have been moved to the `packages/runtime/container-runtime/src/test/batching.spec.ts`. The tests validate the same scenarios with updated patterns to validate them.
- There was a delta manager test in `packages/test/functional-tests/src/test/containerRuntime.spec.ts` which has been moved to delta manager unit tests.

[AB#13509](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/13509)

### CODE REVIEW

File: packages/runtime/container-runtime/src/test/batching.spec.ts [hunk 1]
```diff
@@ -138,106 +146,249 @@ describe("Runtime batching", () => {
...
-		// Change the type of the second message to a non-runtime op.
-		batch[1].type = MessageType.NoOp;
-
-		assert.throws(
-			() => processBatch(batch, containerRuntime),
-			(e: any) => {
-				assert(e.errorType === FluidErrorTypes.dataProcessingError);
-				assert(e.message === "Received a system message during batch processing");
-				return true;
+	describe("Delta scheduler for batches", () => {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/batching.spec.ts:264]: Could you check in these tests in a separate PR like you did with the previous block?

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 2]
```diff
@@ -30,7 +30,7 @@ export class DeltaScheduler {
 	public static readonly processingTime = 50;
 
 	// The increase in time for processing ops after each turn.
-	private readonly processingTimeIncrement = 10;
+	public static processingTimeIncrement = 10;
```
Comment (markfields) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:33]: This should not be changed, right?
```suggestion
	public static readonly processingTimeIncrement = 10;
```

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 3]
```diff
@@ -30,7 +30,7 @@ export class DeltaScheduler {
 	public static readonly processingTime = 50;
 
 	// The increase in time for processing ops after each turn.
```
Comment (markfields) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:32]: Switch to `/** this style */` comment so you can see the comment on hover (consider updating the other fields too)

File: packages/runtime/container-runtime/src/test/batching.spec.ts [hunk 4]
```diff
@@ -138,106 +146,249 @@ describe("Runtime batching", () => {
...
+			// Add delay such that each op takes greater than the DeltaScheduler's processing time to process.
+			// The times increases every time a batch is processed so simulate by increasing the delay.
+			clock.tick(DeltaScheduler.processingTime + deltaSchedulerTimeIncrement);
+			deltaSchedulerTimeIncrement += DeltaScheduler.processingTimeIncrement;
+		}
+
+		let containerRuntimeStub: sinon.SinonStub;
+		let batchBeginCount = 0;
+		let batchEndCount = 0;
+		let deltaSchedulerTimeIncrement = DeltaScheduler.processingTimeIncrement;
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/batching.spec.ts:276]: I'd not name this "increment", it's the sum of N increments.  So maybe "buffer" or something

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 5]
```diff
@@ -1764,12 +1765,7 @@ export class ContainerRuntime
 			closeContainer: (error?: ICriticalContainerError) => this.closeFn(error),
 		});
 
-		this.scheduleManager = new ScheduleManager(
-			this.innerDeltaManager,
-			this,
-			() => this.clientId,
-			createChildLogger({ logger: this.logger, namespace: "ScheduleManager" }),
-		);
+		this.deltaScheduler = new DeltaScheduler(this.innerDeltaManager, this.logger);
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1768]: Want a namespace for the logger?

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 6]
```diff
@@ -2680,46 +2676,80 @@ export class ContainerRuntime
...
-						local,
-						isRuntimeMessage: true,
-						savedOp,
-						localOpMetadata,
-					};
-					this.ensureNoDataModelChanges(() => this.processRuntimeMessage(msg));
-				});
+				this.batchBegin(messagesWithPendingState[0]?.message);
+
+				let error: any;
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2681]: `error: unknown` is preferred (and update the type where it's used)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 7]
```diff
@@ -2680,46 +2676,80 @@ export class ContainerRuntime
 				local,
 			);
 			if (messagesWithPendingState.length > 0) {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2678]: I think this code might read better if you pulled everything in this `if` block into a function `processBatch` (same signature as `processEmptyBatch`), and then in-lined `processRuntimeMessage` (or not, I guess).  It keeps this function more concise and focused on the overall flow of handling unknown messages, grouping into batches, and dispatching.

I guess you'd want a third function `processSingleMessage` for the `else` case below...?  Maybe there's an even clearer way to refactor it that's not jumping out at me.

As-is though this function is doing a lot of different things and would be nice to refactor somewhat.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 8]
```diff
@@ -2680,46 +2676,80 @@ export class ContainerRuntime
...
+					error = e;
+					throw e;
+				} finally {
+					this.batchEnd(
+						error,
+						messagesWithPendingState[messagesWithPendingState.length - 1]?.message,
+					);
+				}
 			} else {
 				this.ensureNoDataModelChanges(() => this.processEmptyBatch(inboundBatch, local));
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2703]: What are your thoughts on why `processEmptyBatch` doesn't call delta scheduler?  To me, it is similar to the `else` case below (legacy or non-runtime message) - we got a single message over the wire (the empty grouped batch, in this case).  So if we notify in that case we should do the same here.

It looks like DeltaScheduler only needs to know the sequenceNumber, so you can just change `message: ISequencedDocumentMessage` in `batchBegin`/`batchEnd` sigature to `{ sequenceNumber }: { sequenceNumber: number }`, similar to what `processEmptyBatch` uses for the event args.

Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2703]: I guess it's been that way since @dannimad introduced `processEmptyBatch` - Daniel any input here?  IMO these different codepaths should be as similar as possible (and share as much of that code as possible too)

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1]: I'm thinking about the test matrix here to cover all the code you changed:

* 4 different types of incoming message (modern runtime message, empty grouped batch, legacy runtime message, non-runtime message)
* Whether or not process throws

In all 8 combos, we want to be sure that "batchBegin/End" event fires and DeltaScheduler is likewise notified.

Up to you if you get that thorough, but wanted to share the concept.  If you can figure out how to refactor to share more code that helps too.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 9]
```diff
@@ -2680,46 +2676,80 @@ export class ContainerRuntime
...
+				throw e;
+			} finally {
+				this.batchEnd(error, messageCopy);
 			}
 		}
 	}
 
+	private batchBegin(message?: ISequencedDocumentMessage) {
+		assert(message !== undefined, "Batch must have a message");
+		this.deltaScheduler.batchBegin(message);
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2743]: What if the deltaScheduler listen for the "batchBegin" and "batchEnd" events?

As it is, the DeltaScheduler just sits there and does what it's told, and ContainerRuntime is the one who "knows" that it needs the DeltaScheduler to do something.  If we flip it, then the ContainerRuntime has less knowledge/opinion, and the DeltaScheduler is more independent, handling the events whenever they come.

I lean towards switching to event-based, it feels more modular - the ContainerRuntime doesn't bake in knowledge about the DeltaScheduler besides creating it.  We don't even need the class field anymore (but note that DeltaScheduler will need to listen for "dispose" event and dispose itself to release the event listeners)

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 8 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2703]: See comment below about suggesting to remove the explicit DeltaScheduler call and have that listen to the events.

Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2703]: Also, FYI the empty batch case is not respecting the event args type specified on `IContainerRuntimeBaseEvents`.  The alternative would be to pass the original empty grouped batch (simply `messageCopy`, I believe).  cc @dannimad for this as well.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 7 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2678]: I agree that this code could do with some refactoring. However, I would prefer to do that in a separate PR because it will make the diff much harder to review. There are other things here like `processRuntimeMessage` and `observeNonRuntimeMessage` having lot of duplicate code that could be cleaned.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2703]: I am not too sure about the empty batch case. It's incorrect that it violates the `IContainerRuntimeBaseEvents` arguments. Anyone listening to this can fail unexpectedly. I would prefer not to emit these events if there was nothing processed.
I would not want to change it in this PR because that is significant change in behavior and I don't know what it's supposed to do.
@dannimad can you please look into this.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 9 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2743]: That sounds fine to me. However, I do not want to change that behavior in this PR. It can be done as a follow up. It will lead to other behavior change where this will be called for an empty batch which isn't a big deal but would be nice to do it separately.

File: packages/runtime/container-runtime/src/test/batching.spec.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/batching.spec.ts:264]: This needs the change I am making in this PR. So, I could do this as a follow up and skip the existing tests for now. IMO, having it part of this PR makes sense but if it helps with review, I could do a follow up.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 8 above)
Comment (dannimad) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2703]: I believe we should call delta scheduler on the empty batch case as well. We could just make slight changes to current container::batchBegin and batchEnd methods and call them from processEmptyBatch(). 

These side effects were implanted in the empty batch case just to make sure the workflow were as close as the non-empty case as possible, not because we must but out of ignorance to know what could be some unwanted consequences of making that case more unique.

Comment (dannimad) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2703]: If we have a strong reason not to call it, I would also remove emmiting the event from the empty case (which does not sound right to me).

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1]: Added tests cases for all except the "process failed on empty batch" because it doesn't process anything so there is nothing to fail.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 7 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: I refactored this - consolidated all non-empty batch processing into a `processBatch` function. That seemed cleaner to me than processing modern vs non-modern messages. Take a look.

**Comment by rajatch-ff** [comment]
> 	public process({ ...messageCopy }: ISequencedDocumentMessage, local: boolean) {

This message technically doesn't process the op anymore, it just constructs the batch as a pre-processing step to `processBatch` method you've introduced. But renaming this will be tough as it's public?

---
Refers to: packages/runtime/container-runtime/src/containerRuntime.ts:2649 in cc04a4b. [](commit_id = cc04a4b9500e162fa93e468613baead816b412aa, deletion_comment = False)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 10]
```diff
@@ -2679,44 +2678,73 @@ export class ContainerRuntime
...
-				);
-			}
+			const batch = [{ message: messageCopy, localOpMetadata: undefined }];
+			this.processBatch(batch, local, savedOp, undefined /* modernRuntimeMessage */);
+		}
+	}
+
+	/**
+	 * Processes a batch of messages. It emits "batchBegin" and "batchEnd" events and calls deltaScheduler.
+	 */
```
Comment (rajatch-ff) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2700]: >*/ [](http://example.com/codeflow?start=2&length=2)

add @param block

**Comment by agarwal-navin** [comment]
> > public process({ ...messageCopy }: ISequencedDocumentMessage, local: boolean) {
> 
> This message technically doesn't process the op anymore, it just constructs the batch as a pre-processing step to `processBatch` method you've introduced. But renaming this will be tough as it's public?
> 
> Refers to: packages/runtime/container-runtime/src/containerRuntime.ts:2649 in [cc04a4b](https://github.com/microsoft/FluidFramework/commit/cc04a4b9500e162fa93e468613baead816b412aa). [](commit_id = [cc04a4b](https://github.com/microsoft/FluidFramework/commit/cc04a4b9500e162fa93e468613baead816b412aa), deletion_comment = False)

Yeah, we need a process method since it's public on `IRuntime`. It can be seen as the core process logic which uses helpers to process different types of messages so it's probably fine..

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 11]
```diff
@@ -2679,44 +2678,73 @@ export class ContainerRuntime
...
-					this.observeNonRuntimeMessage({
-						message: messageCopy as InboundSequencedNonContainerRuntimeMessage,
-						local,
-						isRuntimeMessage: false,
-						savedOp,
-					}),
-				);
-			}
+			const batch = [{ message: messageCopy, localOpMetadata: undefined }];
+			this.processBatch(batch, local, savedOp, undefined /* modernRuntimeMessage */);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2694]: Am I correct, that there is observable change here - we will raise beginBatch & endBatch events for system messages (where we did not in the past)? Is this desired?

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 8 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2690]: I do not have enough context on why we need empty batches (it would be great to have some note somewhere describing why this is needed and how it works), but assuming it's needed, will having runtime noop op (as part of such batch) be a good idea? The metric of "good" for me here - we have less concepts. Partially having noop adds a concept, but not having such things as processEmptyBatch() and special casing here and there might overall reduce number of concepts.

Today, loader will not allow it (if that's good direction - it's easy to change) - would using system noop (i.e. runtime sending "regular" noop) suffice where we leverage today empty batch?

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 12]
```diff
@@ -2818,33 +2834,21 @@ export class ContainerRuntime
 				this.deltaManager.minimumSequenceNumber;
 		}
 
-		// Surround the actual processing of the operation with messages to the schedule manager indicating
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2821]: There is a bunch of duplicated logic between observeNonRuntimeMessage() and processRuntimeMessage().
Maybe created beforeProcessingMessage() & afterProcessingMessage() and move most of the common logic there (to be called from both places)? More chances that if we need to make changes, they will apply to both paths.

File: packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts [hunk 13]
```diff
@@ -200,6 +207,42 @@ export class RemoteMessageProcessor {
 		}
 		assert(batchMetadataFlag !== true, 0x9d6 /* Unexpected batch start marker */);
 
+		// Validate that the sequence numbers are contiguous. If there are non-runtime messages in the batch,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts:210]: Am I correct that we get here only for "legacy" batches?
process() that calls this function has separate code path for isGroupedBatch().

Might be worth updating comments, as it's a bit hard to follow code (and all the combinations).

**Comment by vladsud** [comment]
I believe it's the right direction, but it's worth calling out some risks that are not covered.
The "old" files (files that were not modified for some amount of time) with trailing using "old" batching system will rely on ScheduleManager magic that is gone.
Our document load flow brings all trailing ops (I believe for both ODSP & AFR).
If app (like Loop) uses policy to apply all "cached" ops, then all such ops will be applied to container (in one go) before container is returned to caller, so in essence we do satisfy guarantees of non-breaking batches even without ScheduleManager in such case.

The part where it could break - using defaults and not applying cached ops at once.
The chance of hitting a problem is higher for AzureClient 1.x documents, as such clients were on older version of FF longer.
While chance is really low (it's not just that there needs to be trailing ops in old batch format, it's also that it should matter that they are not applied synchronously), it might be worth considering how we could help users if they were to run into a problem like that (i.e. if we cause a regression).

The only thing that comes to mind - wait for "connected' stated on boot, to ensure that all ops are applied before app moves to rendering / next steps.

**Review by vladsud — APPROVED** [review]

**Comment by agarwal-navin** [comment]
> I believe it's the right direction, but it's worth calling out some risks that are not covered. The "old" files (files that were not modified for some amount of time) with trailing using "old" batching system will rely on ScheduleManager magic that is gone. Our document load flow brings all trailing ops (I believe for both ODSP & AFR). If app (like Loop) uses policy to apply all "cached" ops, then all such ops will be applied to container (in one go) before container is returned to caller, so in essence we do satisfy guarantees of non-breaking batches even without ScheduleManager in such case.
> 
> The part where it could break - using defaults and not applying cached ops at once. The chance of hitting a problem is higher for AzureClient 1.x documents, as such clients were on older version of FF longer. While chance is really low (it's not just that there needs to be trailing ops in old batch format, it's also that it should matter that they are not applied synchronously), it might be worth considering how we could help users if they were to run into a problem like that (i.e. if we cause a regression).
> 
> The only thing that comes to mind - wait for "connected' stated on boot, to ensure that all ops are applied before app moves to rendering / next steps.

Good point about legacy runtime messages. I was able to run the logic that ensures batches are processed together for these messages as well. I added tests and validated that it worked.
However, when I run those against existing logic (without this PR), it fails in the below code with assert `0x29b`:
```
if (!isRuntimeMessage(message)) {
	// Protocol messages should never show up in the middle of the batch!
	if (this.currentBatchClientId !== undefined) {
		throw DataProcessingError.create(
			"Received a system message during batch processing", // Formerly known as assert 0x29a
			"trackPending",
			message,
			{
				runtimeVersion: pkgVersion,
				batchClientId:
					// eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
					this.currentBatchClientId === null ? "null" : this.currentBatchClientId,
				pauseSequenceNumber: this.pauseSequenceNumber,
				localBatch: this.currentBatchClientId === this.getClientId(),
				messageType: message.type,
			},
		);
	}

	assert(batchMetadata === undefined, 0x29b /* "system op in a batch?" */);
	return;
}
```

The `isRuntimeMessage` only checks whether the type is `MessageType.Operation` or not. So, for messages with type like `ContainerMessageType.FluidDataStoreOp`, it fails. So, the current code doesn't work for these types of messages. I could fix it, but should I? Considering it doesn't even work.

File: packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts (hunk 13 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts:210]: Yes, that is true. We have too many combinations :-). I realized that this logic is only for non-grouped ops. There is no way that batches for grouped ops could be incorrect so that's why there is no validation and I kept it that way.

I will add comments to make it clear.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 12 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2821]: Yes, there is. We are in the middle of refactoring this code I believe. I don't want to change it in this PR. I will consider doing a follow up to clean this up more. Mark also has some comments about refactoring and I can do a refactoring PR later.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 8 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2690]: We considered adding a runtime no-op, but didn't wnat to wait for saturation blah blah blah.  So we went with an empty grouped batch.  I think we should just project that empty grouped batch op to the rest of the system, no one should be inspecting it anyway.  It would be surprising to anyone who _was_ closely inspecting the op payload, because other than this you'd never expect to see a Grouped Batch at this layer (it would be ungrouped and you'd get actual batch messages themselves in these cases).  But I think it's fine.

Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2690]: Several people have asked for info about why empty batches.  There are some comments scattered around wherever we are dealing with empty batches, but that's not enough.

Where did you look for info but not find it?  I'll add it there.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 14]
```diff
@@ -2679,44 +2678,73 @@ export class ContainerRuntime
...
+	) {
+		const firstMessage = batch[0]?.message;
+		assert(firstMessage !== undefined, "Batch must have at least one message");
+		this.deltaScheduler.batchBegin(firstMessage);
+		this.emit("batchBegin", firstMessage);
+
+		let error: unknown;
+		try {
+			batch.forEach(({ message, localOpMetadata }) => {
+				const runtimeMessage = modernRuntimeMessage ?? isRuntimeMessage(message);
```
Comment (markfields) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2718]: nit: make `modernRuntimeMessage` a boolean and then 
```suggestion
				const runtimeMessage = modernRuntimeMessage || isRuntimeMessage(message);
```

Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2718]: (Optional) Also, a thought - this code makes it seem like each message in the batch needs to check if it's a runtime message, as if you could have a mixed batch.  I understand the shared code and how some callers will pass a singleton array.  But someone else wouldn't.

Maybe move the runtimeMessage calculation outside the loop, something like this:

```typescript
const runtimeMessage = modernRuntimeMessage || batch.every(({ message }) => isRuntimeMessage(message));
assert(!runtimeMessage || batch.every(({ message }) => !isRuntimeMessage(message)), "Batches should be homogenous regarding runtime message or not");
```

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 12 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2821]: @pragya91 split them apart but left them (nearly) identical for the express purpose of having them be separate so we could carefully remove functionality from the `observe` function.  So let's not combine them unless we have changed our mind about diverging treatment of these two groups.

File: packages/runtime/container-runtime/src/test/batching.spec.ts (hunk 1 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/test/batching.spec.ts:264]: Nah it's fine, it's just always reassurring the have the tests there and not changing (and still passing) when making a refactor like this.

File: packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts [hunk 15]
```diff
@@ -200,6 +207,42 @@ export class RemoteMessageProcessor {
...
+		// the remote message processor isn't called and so it will appear like the batch is missing messages
+		// and this case will be hit.
+		const previousMessage =
+			this.batchInProgress.messages[this.batchInProgress.messages.length - 1];
+		if (
+			previousMessage !== undefined &&
```
Comment (markfields) [inline packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts:216]: Looks to me like you can (and should) assert that previousMessage is not undefined

**Comment by markfields** [comment]
> The isRuntimeMessage only checks whether the type is MessageType.Operation or not. So, for messages with type like ContainerMessageType.FluidDataStoreOp, it fails
Basically, the current code doesn't work for these types of messages. I could fix it, but should I? Considering it doesn't even work.

@agarwal-navin Can you at least open a bug that describes the exact repro case?  I think it might be a quick fix (add another check on the message type), so we should evaluate the risk/benefit there, we can discuss in the bug.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 8 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2690]: Opened AB#13989 about the gaps around empty batches

**Comment by agarwal-navin** [comment]
Putting this on hold for now because it will break the last seq# and min seq# tracking in delta manager.
In the mean time, I have another PR that simplfy schedule manager and the process pipeline- https://github.com/microsoft/FluidFramework/pull/22356.
